### ISSUE TITLE
Another reason to not allow third party cookies

### ISSUE THREAD

**Issue by Gitoffthelawn** [ghacksuserjs/ghacks-user.js#159]
IndexedDB uses the same-origin principle, which means that it ties the store to the origin of the site that creates it (typically, this is the site domain or subdomain), so it cannot be accessed by any other origin.

**Third party window content (e.g. content) can access the IndexedDB store for the origin it is embedded into, unless the browser is set to [never accept third party cookies](https://support.mozilla.org/kb/disable-third-party-cookies) (see bug [1147821](https://bugzilla.mozilla.org/show_bug.cgi?id=1147821)),**

Reference: https://developer.mozilla.org/docs/Web/API/IndexedDB_API/Using_IndexedDB#Security

Thanks to: https://github.com/ChrisAntaki/disable-indexeddb/issues/1#issuecomment-311381692

**Comment by Thorin-Oakenpants** [comment]
https://www.fxsitecompat.com/en-CA/docs/2015/web-storage-indexeddb-cache-api-now-obey-third-party-cookies-preference/

This is not just IndexedDB, it's Cache API, Web Storage and Service Worker Cache. I think we could add a [NOTE] to 2701. :+1: or nits

```js
/* 2701: disable cookies on all sites [SETUP]
 * blah blah
 * [NOTE] This also controls 3rd party Web Storage, IndexedDB, Cache API and Service Worker Cache access
 * [1] https://www.fxsitecompat.com/en-CA/docs/2015/web-storage-indexeddb-cache-api-now-obey-third-party-cookies-preference/ ***/
```

I know cookies and local dom have some sort of connection (see @RoxKilly 's issue with material angle io website) - but to be honest I was not sure if that was Cookie Controller or FF permissions. I never realized it controlled indexedDB etc in some fashion. I wonder if FPI covers indexedDB.

**Comment by Gitoffthelawn** [comment]
@Thorin-Oakenpants  What is FPI?

**Comment by Thorin-Oakenpants** [comment]
First Party Isolation - pref 2698 ... the one we have a sticky issue for, part of the Tor Uplift

**Comment by Gitoffthelawn** [comment]
Thanks

**Comment by Thorin-Oakenpants** [comment]
https://www.wilderssecurity.com/threads/firefox-and-indexeddb.395113/

current js content
```js
/* 2404: disable JS storing data permanently
 * This setting WAS under about:permissions>All Sites>Maintain Offline Storage
 * [NOTE] about:permissions is no longer available since FF46 but you can still override
 * individual domains: use info icon in urlbar etc or right click on a web page>view page info
 * [WARNING] [SETUP] If set as false (disabled), this WILL break some [old] add-ons and DOES
 * break a lot of sites' functionality. Applies to websites, add-ons and session data.
 * [1] https://addons.mozilla.org/en-US/firefox/addon/disable-indexeddb/ ***/
user_pref("dom.indexedDB.enabled", false);
```

The problem - indexedDB is NOT controlled by site exceptions: i.e
- urlbar info icon > right arrow > more information > permissions > maintain offline storage
- on a site such as google or the new york times, an icon will appear in the urlbar after the info icon, which you can hover over and it will say "You have blocked data storage for this website" - click it and you an option to unblock it.

Can anyone confirm this for me, to confirm the original linked OPs test on New York Times (or use some other site). We will need to fix the description if this is the case.

---
I also used the second method above on google.com, clicked the icon, unblocked and it said to reload the page, so I did, and the icon did not reappear. So far so good. But the permissions tab (site info etc) is triple state (default, ask, block). So to reset it, I went in there, but it was already at "default" - and yet no icon appeared. So I set it to "block" and now it does appear. Now I have to use a site override (block) whereas before it was always default and google always had the icon come up. If I reset it to default, now I never get the icon. F**k. What the hell is going on.

**Comment by Thorin-Oakenpants** [comment]
What is this controlling? urlbar info icon > right arrow > more information > permissions > maintain offline storage

My default is "always ask". I have played with some settings. AFAIK, it's not
- `dom.storage.enabled` (2705)
- `dom.storageManager.enabled` (2706)
- `dom.indexedDB.enabled` (2404) according to summerheat at the wilders security post. but my gut says this is what it is meant to control, or used to control. But my pref is "false", and yet my default in the permissions tab is "always ask"

What am I missing here. What is the pref that changes "Always ask" to "block" - that way I can have a default deny-all and assign site exceptions

**Comment by earthlng** [comment]
> What is the pref that changes "Always ask" to "block" - that way I can have a default deny-all and assign site exceptions

https://dxr.mozilla.org/mozilla-central/source/browser/locales/en-US/chrome/browser/browser.dtd#232
-> it should definitely be "indexedDB" if it says "You have blocked data storage for this website" as you wrote above
https://dxr.mozilla.org/mozilla-central/source/browser/modules/SitePermissions.jsm#532
-> looks like not all permissions can be set to "block" by default. Currently (nightly) it's only possible for image, cookie, popup and install. For me everything except popup and install is either defaulting to "allow" or "always ask" (I allow cookies and control them with uMatrix - everything else should match what you get) Maybe there's other code somewhere that I can't find that also looks at other prefs for that GUI but I think that's largely it.

**Comment by Thorin-Oakenpants** [comment]
@Gitoffthelawn wrote in the addons sticky and now needs to delete it to help keep that thread clean

> Thorin-Smokinpants wrote:
> 
> > Block all cookies by default (esp 3rd party as this has other ramifications with access to indexedDB, cache etc).
> 
> I previously created an issue (https://github.com/ghacksuserjs/ghacks-user.js/issues/159) regarding the ramifications with indexedDB.  What are the cache ramifications?

**Comment by Thorin-Oakenpants** [comment]
I have no idea regards "cache" - by that I assume you mean being able to clean indexedDB?

I actually think this Maintain Offline Storage pref is broken in some way. @zymase and I discussed something about this back at ghacks ages ago. I can't remember or find exactly what it was about, might have been dom storage or indexedDB. I think I may have seen some bugzilla tickets on it.

**Comment by Gitoffthelawn** [comment]
> I have no idea regards "cache" - by that I assume you mean being able to clean indexedDB?

I was referring to "cache" in:

> Block all cookies by default (esp 3rd party as this has other ramifications with access to indexedDB, ***cache*** etc).

**Comment by Thorin-Oakenpants** [comment]
the ink in the second post (https://www.fxsitecompat.com/en-CA/docs/2015/web-storage-indexeddb-cache-api-now-obey-third-party-cookies-preference/ ) .. it lists `Cache API` and `Service Worker Cache` among the items

**Comment by Gitoffthelawn** [comment]
Thanks.  He's implying that third party sites are not cached if third-party cookies are disabled.... yet the content is still loaded.  Interesting.

**Comment by Thorin-Oakenpants** [comment]
Who is implying? The fxsitecompat article? Article links to https://bugzilla.mozilla.org/show_bug.cgi?id=1145744

> The Cache API should not be accessible in Window or workers for 3rd party iframes when the user has disable 3rd party cookies

This has nothing to do with it being downloaded, it is all about **accessibility** to to from 3rd parties AFAIK

**Comment by Gitoffthelawn** [comment]
I made an edit to be clear... sorry for the ambiguous words.

That's great that the cache API is not available when third-party cookies are disabled.

Yet another reason to disallow them.

Note that blocking with other tools (like uMatrix) will not have this benefit.

**Comment by Thorin-Oakenpants** [comment]
re-closing this - started a clean new issue #173
